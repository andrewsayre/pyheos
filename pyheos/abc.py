"""Define abstract base classes for HEOS."""

from abc import ABC
from typing import Any


class RemoveHeosFieldABC(ABC):
    """Define an abstract base class that removes the 'heos' from dataclass's fields list to prevent serialization."""

    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
        """Post initialize the player."""
        # Prevent the heos instance from being serialized
        fields = self.__dataclass_fields__.copy()  # type: ignore[has-type] # pylint: disable=access-member-before-definition
        del fields["heos"]
        self.__dataclass_fields__ = fields
