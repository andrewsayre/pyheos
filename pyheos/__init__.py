"""pyheos - a library for interacting with HEOS devices."""
from . import const
from .dispatch import Dispatcher
from .heos import Heos
from .player import HeosNowPlayingMedia, HeosPlayer
from .response import CommandError
from .source import HeosSource, InputSource

__all__ = [
    'const',
    'CommandError',
    'Dispatcher',
    'Heos',
    'HeosPlayer',
    'HeosNowPlayingMedia',
    'HeosSource',
    'InputSource'
]
