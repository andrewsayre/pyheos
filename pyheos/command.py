"""Define the HEOS command module."""
from typing import Sequence, Tuple

from . import const
from .player import HeosNowPlayingMedia


class HeosCommands:
    """Define a class that encapsulates well-known commands."""

    def __init__(self, connection):
        """Init the command wrapper."""
        self._connection = connection

    async def heart_beat(self, *, raise_for_result=False) -> bool:
        """Perform heart beat command."""
        response = await self._connection.command(
            const.COMMAND_HEART_BEAT, raise_for_result,
            raise_for_result=raise_for_result)
        return response.result

    async def register_for_change_events(
            self, enable=True, *, raise_for_result=False) -> bool:
        """Enable or disable change event notifications."""
        params = {
            'enable': "on" if enable else "off"
        }
        response = await self._connection.command(
            const.COMMAND_REGISTER_FOR_CHANGE_EVENTS, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_players(self) -> Sequence[dict]:
        """Get players."""
        response = await self._connection.command(
            const.COMMAND_GET_PLAYERS, raise_for_result=True)
        return response.payload

    async def get_player_state(self, player_id: int) -> str:
        """Get the state of the player."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_PLAY_STATE, params, raise_for_result=True)
        return response.get_message('state')

    async def set_player_state(self, player_id: int, state: str,
                               *, raise_for_result=False) -> bool:
        """Set the state of the player."""
        if state not in const.VALID_PLAY_STATES:
            raise ValueError("Invalid play state: " + state)
        params = {
            'pid': player_id,
            'state': state
        }
        response = await self._connection.command(
            const.COMMAND_SET_PLAY_STATE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_now_playing_state(
            self, player_id: int,
            now_playing_media: HeosNowPlayingMedia):
        """Get the now playing media information."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_NOW_PLAYING_MEDIA, params, raise_for_result=True)
        now_playing_media.from_data(response.payload)

    async def get_volume(self, player_id: int) -> int:
        """Get the volume of the player."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_VOLUME, params, raise_for_result=True)
        return int(response.get_message('level'))

    async def set_volume(self, player_id: int, level: int,
                         *, raise_for_result=False) -> bool:
        """Set the volume of the player."""
        if level < 0 or level > 100:
            raise ValueError("'level' must be in the range 0-100")
        params = {
            'pid': player_id,
            'level': level
        }
        response = await self._connection.command(
            const.COMMAND_SET_VOLUME, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_mute(self, player_id: str) -> bool:
        """Get the mute state of the player."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_MUTE, params, raise_for_result=True)
        return response.get_message('state') == 'on'

    async def set_mute(self, player_id: str, state: bool,
                       *, raise_for_result=False) -> bool:
        """Set the mute state of the player."""
        params = {
            'pid': player_id,
            'state': "on" if state else "off"
        }
        response = await self._connection.command(
            const.COMMAND_SET_MUTE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def volume_up(self, player_id: int,
                        step: int = const.DEFAULT_STEP,
                        *, raise_for_result=False) -> bool:
        """Increase the volume level."""
        if step < 1 or step > 10:
            raise ValueError("'step' must be in the range 1-10")
        params = {
            'pid': player_id,
            'step': step
        }
        response = await self._connection.command(
            const.COMMAND_VOLUME_UP, params,
            raise_for_result=raise_for_result)
        return response.result

    async def volume_down(self, player_id: int,
                          step: int = const.DEFAULT_STEP,
                          *, raise_for_result=False) -> bool:
        """Increase the volume level."""
        if step < 1 or step > 10:
            raise ValueError("'step' must be in the range 1-10")
        params = {
            'pid': player_id,
            'step': step
        }
        response = await self._connection.command(
            const.COMMAND_VOLUME_DOWN, params,
            raise_for_result=raise_for_result)
        return response.result

    async def toggle_mute(self, player_id: int,
                          *, raise_for_result=False) -> bool:
        """Toggle the mute state.."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_TOGGLE_MUTE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_play_mode(self, player_id: int) -> Tuple[str, bool]:
        """Get the current play mode."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_GET_PLAY_MODE, params, raise_for_result=True)
        repeat = response.get_message('repeat')
        shuffle = response.get_message('shuffle') == 'on'
        return repeat, shuffle

    async def set_play_mode(self, player_id: int, repeat: str,
                            shuffle: bool, *, raise_for_result=False):
        """Set the current play mode."""
        if repeat not in const.VALID_REPEAT_MODES:
            raise ValueError("Invalid repeat mode: " + repeat)
        params = {
            'pid': player_id,
            'repeat': repeat,
            'shuffle': 'on' if shuffle else 'off'
        }
        response = await self._connection.command(
            const.COMMAND_SET_PLAY_MODE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def clear_queue(self, player_id: int,
                          *, raise_for_result=False) -> bool:
        """Clear the queue."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_CLEAR_QUEUE, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_next(self, player_id: int,
                        *, raise_for_result=False) -> bool:
        """Play next."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_PLAY_NEXT, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_previous(self, player_id: int,
                            *, raise_for_result=False) -> bool:
        """Play next."""
        params = {
            'pid': player_id
        }
        response = await self._connection.command(
            const.COMMAND_PLAY_PREVIOUS, params,
            raise_for_result=raise_for_result)
        return response.result

    async def get_music_sources(self) -> Sequence[dict]:
        """Get available music sources."""
        response = await self._connection.command(
            const.COMMAND_BROWSE_GET_SOURCES, raise_for_result=True)
        return response.payload

    async def browse(self, source_id: int) -> Sequence[dict]:
        """Browse a music source."""
        params = {
            'sid': source_id
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_BROWSE, params, raise_for_result=True)
        return response.payload

    async def play_input(self, player_id: int, input_name: str, *,
                         source_player_id: int = None,
                         raise_for_result=False) -> bool:
        """Play the specified input source."""
        if input_name not in const.VALID_INPUTS:
            raise ValueError("Invalid input name: " + input_name)
        params = {
            'pid': player_id,
            'spid': source_player_id or player_id,
            'input': input_name
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_PLAY_INPUT, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_preset(self, player_id: int, preset: int,
                          *, raise_for_result=False) -> bool:
        """Play the specified preset by 1-based index."""
        if preset < 1:
            raise ValueError("Invalid preset: " + str(preset))
        params = {
            'pid': player_id,
            'preset': preset
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_PLAY_PRESET, params,
            raise_for_result=raise_for_result)
        return response.result

    async def play_stream(self, player_id: int, url: str,
                          *, raise_for_result=False) -> bool:
        """Play the specified URL."""
        params = {
            'pid': player_id,
            'url': url
        }
        response = await self._connection.command(
            const.COMMAND_BROWSE_PLAY_STREAM, params,
            raise_for_result=raise_for_result)
        return response.result
