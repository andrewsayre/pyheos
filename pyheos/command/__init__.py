"""Define the HEOS command module."""

from typing import Final

# Browse commands
COMMAND_BROWSE_GET_SOURCES: Final = "browse/get_music_sources"
COMMAND_BROWSE_BROWSE: Final = "browse/browse"
COMMAND_BROWSE_PLAY_INPUT: Final = "browse/play_input"
COMMAND_BROWSE_PLAY_PRESET: Final = "browse/play_preset"
COMMAND_BROWSE_PLAY_STREAM: Final = "browse/play_stream"
COMMAND_BROWSE_ADD_TO_QUEUE: Final = "browse/add_to_queue"

# Player commands
COMMAND_GET_PLAYERS: Final = "player/get_players"
COMMAND_GET_PLAY_STATE: Final = "player/get_play_state"
COMMAND_SET_PLAY_STATE: Final = "player/set_play_state"
COMMAND_GET_NOW_PLAYING_MEDIA: Final = "player/get_now_playing_media"
COMMAND_GET_VOLUME: Final = "player/get_volume"
COMMAND_SET_VOLUME: Final = "player/set_volume"
COMMAND_GET_MUTE: Final = "player/get_mute"
COMMAND_SET_MUTE: Final = "player/set_mute"
COMMAND_VOLUME_UP: Final = "player/volume_up"
COMMAND_VOLUME_DOWN: Final = "player/volume_down"
COMMAND_TOGGLE_MUTE: Final = "player/toggle_mute"
COMMAND_GET_PLAY_MODE: Final = "player/get_play_mode"
COMMAND_SET_PLAY_MODE: Final = "player/set_play_mode"
COMMAND_CLEAR_QUEUE: Final = "player/clear_queue"
COMMAND_PLAY_NEXT: Final = "player/play_next"
COMMAND_PLAY_PREVIOUS: Final = "player/play_previous"
COMMAND_PLAY_QUICK_SELECT: Final = "player/play_quickselect"
COMMAND_SET_QUICK_SELECT: Final = "player/set_quickselect"
COMMAND_GET_QUICK_SELECTS: Final = "player/get_quickselects"

# Group commands
COMMAND_GET_GROUPS: Final = "group/get_groups"
COMMAND_SET_GROUP: Final = "group/set_group"
COMMAND_GET_GROUP_VOLUME: Final = "group/get_volume"
COMMAND_SET_GROUP_VOLUME: Final = "group/set_volume"
COMMAND_GET_GROUP_MUTE: Final = "group/get_mute"
COMMAND_SET_GROUP_MUTE: Final = "group/set_mute"
COMMAND_GROUP_TOGGLE_MUTE: Final = "group/toggle_mute"
COMMAND_GROUP_VOLUME_UP: Final = "group/volume_up"
COMMAND_GROUP_VOLUME_DOWN: Final = "group/volume_down"

# System commands
COMMAND_REGISTER_FOR_CHANGE_EVENTS: Final = "system/register_for_change_events"
COMMAND_HEART_BEAT: Final = "system/heart_beat"
COMMAND_ACCOUNT_CHECK: Final = "system/check_account"
COMMAND_REBOOT: Final = "system/reboot"
COMMAND_SIGN_IN: Final = "system/sign_in"
COMMAND_SIGN_OUT: Final = "system/sign_out"
