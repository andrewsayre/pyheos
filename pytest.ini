[pytest]
log_level=DEBUG
testpaths = tests
norecursedirs = .git
timeout = 60
asyncio_default_fixture_loop_scope = function