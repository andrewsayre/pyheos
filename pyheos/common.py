"""Define the common module."""

from typing import cast

from pyheos.const import TARGET_VERSION


def is_supported_version(
    version: str | None, min_version: str = TARGET_VERSION
) -> bool:
    """Check if a version is supported.

    Args:
        version (str): The version to check.
        min_version (str): The minimum version.

    Returns:
        bool: True if the version is supported, False otherwise.
    """
    if version is None:
        return False
    try:
        sem_ver = get_sem_ver(version)
    except ValueError:
        return False
    min_sem_ver = get_sem_ver(min_version)
    for i, ver in enumerate(sem_ver):
        if ver < min_sem_ver[i]:
            return False
        if ver > min_sem_ver[i]:
            return True
    return True


def get_sem_ver(version: str) -> tuple[int, int, int]:
    """Get the semantic version from a string.

    Args:
        version (str): The version string.

    Returns:
        tuple: The semantic version as a tuple.
    """
    return cast(tuple[int, int, int], tuple(map(int, version.split(".", 3))))
