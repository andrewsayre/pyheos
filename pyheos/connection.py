"""Define the connection module."""

import asyncio
from collections import defaultdict
from datetime import datetime, timedelta
import json
import logging
from typing import Any, DefaultDict, Dict, List, Optional

from . import const
from .command import HeosCommands
from .response import HeosResponse

SEPARATOR = '\r\n'
SEPARATOR_BYTES = SEPARATOR.encode()

_LOGGER = logging.getLogger(__name__)


_QUOTE_MAP = {
    '&': '%26',
    '=': '%3D',
    '%': '%25'
}


def _quote(string: str) -> str:
    """Quote a string per the CLI specification."""
    return ''.join([_QUOTE_MAP.get(char, char) for char in str(string)])


def _encode_query(items: dict) -> str:
    """Encode a dict to query string per CLI specifications."""
    pairs = []
    for key, value in items.items():
        item = key + "=" + _quote(value)
        # Ensure 'url' goes last per CLI spec
        if key == 'url':
            pairs.append(item)
        else:
            pairs.insert(0, item)
    return '&'.join(pairs)


class HeosConnection:
    """Define a class that encapsulates read/write."""

    def __init__(self, heos, host: str, *,
                 timeout: float = const.DEFAULT_TIMEOUT,
                 heart_beat: Optional[float] = const.DEFAULT_HEART_BEAT,
                 all_progress_events=True):
        """Init a new HeosConnection class."""
        self._heos = heos
        self._all_progress_events = all_progress_events
        self.host = host  # type: str
        self.commands = HeosCommands(self)
        self.timeout = timeout  # type: int
        self._reader = None  # type: asyncio.StreamReader
        self._writer = None   # type: asyncio.StreamWriter
        self._response_handler_task = None  # type: asyncio.Task
        self._pending_commands = \
            defaultdict(list)  # type: DefaultDict[str, List[ResponseEvent]]
        self._sequence = 0  # type: int
        self._state = const.STATE_DISCONNECTED  # type: str
        self._auto_reconnect = False  # type: bool
        self._reconnect_delay = const.DEFAULT_RECONNECT_DELAY  # type: float
        self._reconnect_task = None  # type: asyncio.Task
        self._last_activity = None  # type: datetime
        self._heart_beat_interval = heart_beat  # type: Optional[float]
        self._heart_beat_task = None  # type: asyncio.Task

    async def connect(self, *, auto_reconnect: bool = False,
                      reconnect_delay: float = const.DEFAULT_RECONNECT_DELAY):
        """Invoke the connect operation."""
        if self._state == const.STATE_CONNECTED:
            return
        # Ensure we don't try to reconnect during initial failures
        self._auto_reconnect = False
        await self._connect()
        self._auto_reconnect = auto_reconnect
        self._reconnect_delay = reconnect_delay

    async def _connect(self):
        """Perform core connection logic."""
        open_future = asyncio.open_connection(
            self.host, const.CLI_PORT)
        self._reader, self._writer = await asyncio.wait_for(
            open_future, self.timeout)
        # Start response handler
        self._response_handler_task = asyncio.ensure_future(
            self._response_handler())
        # Set state before calling command as it checks for handling
        self._state = const.STATE_CONNECTED
        await self.commands.register_for_change_events()
        # Start heart beat if enabled.
        if self._heart_beat_interval is not None \
                and self._heart_beat_interval > 0:
            self._heart_beat_task = asyncio.ensure_future(
                self._heart_beat())

        _LOGGER.debug("Connected to %s", self.host)
        self._heos.dispatcher.send(
            const.SIGNAL_HEOS_EVENT, const.EVENT_CONNECTED)

    async def disconnect(self):
        """Disconnect from the device."""
        if self._state == const.STATE_DISCONNECTED:
            return
        # Cancel pending reconnect
        if self._reconnect_task:
            self._reconnect_task.cancel()
            await self._reconnect_task
            self._reconnect_task = None
        # Core disconnect
        await self._disconnect()
        self._state = const.STATE_DISCONNECTED

        _LOGGER.debug("Disconnected from %s", self.host)
        self._heos.dispatcher.send(
            const.SIGNAL_HEOS_EVENT, const.EVENT_DISCONNECTED)

    async def _disconnect(self):
        # Cancel response handler
        if self._heart_beat_task:
            self._heart_beat_task.cancel()
            self._heart_beat_task = None
        if self._response_handler_task:
            self._response_handler_task.cancel()
            await self._response_handler_task
            self._response_handler_task = None
        # Close channel
        if self._writer:
            self._writer.close()
            self._writer = None
        self._reader = None
        self._sequence = 0
        self._pending_commands.clear()

    async def _handle_connection_error(self, error: Exception):
        """Handle connection failures and schedule reconnect."""
        await self._disconnect()

        if self._auto_reconnect:
            self._state = const.STATE_RECONNECTING
            self._reconnect_task = asyncio.ensure_future(self._reconnect())
        else:
            self._state = const.STATE_DISCONNECTED

        _LOGGER.debug("Disconnected from %s", self.host, exc_info=error)
        self._heos.dispatcher.send(
            const.SIGNAL_HEOS_EVENT, const.EVENT_DISCONNECTED)

    async def _reconnect(self):
        """Perform core reconnection logic."""
        while self._state != const.STATE_CONNECTED:
            try:
                await self._connect()
                self._reconnect_task = None
                return
            except (ConnectionError, asyncio.TimeoutError):
                # Occurs when we could not reconnect
                _LOGGER.debug("Failed to reconnect to %s", self.host,
                              exc_info=True)
                await self._disconnect()
                await asyncio.sleep(self._reconnect_delay)
            except asyncio.CancelledError:
                # Occurs when reconnect is cancelled via disconnect
                return

    async def _response_handler(self):
        while True:
            # Wait for response
            try:
                result = await self._reader.readuntil(SEPARATOR_BYTES)
                self._last_activity = datetime.utcnow()
                data = json.loads(result.decode())
                response = HeosResponse(data)

                # Ignore processing
                if response.is_under_process:
                    _LOGGER.debug("Command under process '%s': '%s'",
                                  response.command, data)
                    continue
                # Handle events
                if response.is_event:
                    asyncio.ensure_future(self._handle_event(response))
                    continue
                # Find pending command
                commands = self._pending_commands[response.command]
                if not commands:
                    _LOGGER.debug("Received response with no pending "
                                  "command: '%s'", response.command)
                    continue
                sequence = response.get_message('sequence')
                if sequence:
                    sequence_id = int(sequence)
                    event = next((event for event in commands
                                  if event.sequence == sequence_id), None)
                    commands.remove(event)
                else:
                    event = commands.pop(0)
                event.set(response)

            except asyncio.CancelledError:
                # Occurs when the task is being killed
                return
            except (ConnectionError, asyncio.IncompleteReadError) as error:
                # Occurs when the connection breaks
                asyncio.ensure_future(self._handle_connection_error(error))
                return

    async def _heart_beat(self):
        while self._state == const.STATE_CONNECTED:
            try:
                last_activity = datetime.utcnow() - self._last_activity
                threshold = timedelta(seconds=self._heart_beat_interval)
                if last_activity > threshold:
                    asyncio.ensure_future(self.commands.heart_beat())
                await asyncio.sleep(self._heart_beat_interval / 2)
            except asyncio.CancelledError:
                # Occurs when the task is being killed
                return
            except Exception:  # pylint: disable=broad-except
                pass

    async def command(self, command: str,
                      params: Dict[str, Any] = None) -> Optional[HeosResponse]:
        """Run a command and get it's response."""
        if self._state != const.STATE_CONNECTED:
            raise ValueError

        # append sequence number
        sequence = self._sequence
        self._sequence += 1
        params = params or {}
        params['sequence'] = sequence
        command_name = command
        uri = const.BASE_URI + command + '?' + _encode_query(params)

        # Add reservation
        event = ResponseEvent(sequence)
        pending_commands = self._pending_commands[command_name]
        pending_commands.append(event)
        # Send command
        self._writer.write((uri + SEPARATOR).encode())
        await self._writer.drain()

        try:
            response = await asyncio.wait_for(event.wait(), self.timeout)
        except (ConnectionError, asyncio.TimeoutError) as error:
            # Occurs when the connection breaks
            asyncio.ensure_future(self._handle_connection_error(error))
            raise

        _LOGGER.debug("Executed command '%s': '%s'", command, response)
        return response

    async def _handle_event(self, response: HeosResponse):
        """Handle a response event."""
        if response.command in const.PLAYER_EVENTS:
            player_id = response.get_player_id()
            player = self._heos.players.get(player_id)
            if player and (await player.event_update(
                    response, self._all_progress_events)):
                self._heos.dispatcher.send(
                    const.SIGNAL_PLAYER_EVENT, player_id, response.command)
                _LOGGER.debug("Event received for player %s: %s",
                              player, response)
        elif response.command in const.GROUP_EVENTS:
            group_id = response.get_group_id()
            self._heos.dispatcher.send(
                const.SIGNAL_GROUP_EVENT, group_id, response.command)
            _LOGGER.debug("Event received for group %s: %s",
                          group_id, response)
        elif response.command in const.HEOS_EVENTS:
            # pylint: disable=protected-access
            result = await self._heos._handle_event(response)
            if result:
                self._heos.dispatcher.send(
                    const.SIGNAL_CONTROLLER_EVENT, response.command)
            _LOGGER.debug("Event received: %s", response)
        else:
            _LOGGER.debug("Unrecognized event: %s", response)

    @property
    def state(self) -> str:
        """Get the current state of the connection."""
        return self._state


class ResponseEvent:
    """Define an awaitable command event response."""

    def __init__(self, sequence: int):
        """Init a new instance of the CommandEvent."""
        self._event = asyncio.Event()
        self._sequence = sequence
        self._response = None

    @property
    def sequence(self) -> int:
        """Get the sequence that represents this event."""
        return self._sequence

    async def wait(self):
        """Wait until the event is set."""
        await self._event.wait()
        return self._response

    def set(self, response: HeosResponse):
        """Set the response."""
        self._response = response
        self._event.set()
