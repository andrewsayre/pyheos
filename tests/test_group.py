"""Tests for the group module."""
import pytest

from pyheos import const


@pytest.mark.asyncio
async def test_set_volume(mock_device, heos):
    """Test the set_volume command."""
    await heos.get_groups()
    group = heos.groups.get(1)
    mock_device.register(const.COMMAND_SET_GROUP_VOLUME,
                         {'level': '25', 'gid': '1'},
                         'group.set_volume')
    with pytest.raises(ValueError):
        await group.set_volume(-1)
    with pytest.raises(ValueError):
        await group.set_volume(101)
    await group.set_volume(25)


@pytest.mark.asyncio
async def test_volume_down(mock_device, heos):
    """Test the volume_down command."""
    await heos.get_groups()
    group = heos.groups.get(1)
    mock_device.register(const.COMMAND_GROUP_VOLUME_DOWN,
                         {'step': '6', 'gid': '1'},
                         'group.volume_down')

    with pytest.raises(ValueError):
        await group.volume_down(0)
    with pytest.raises(ValueError):
        await group.volume_down(11)
    await group.volume_down(6)


@pytest.mark.asyncio
async def test_volume_up(mock_device, heos):
    """Test the volume_up command."""
    await heos.get_groups()
    group = heos.groups.get(1)
    mock_device.register(const.COMMAND_GROUP_VOLUME_UP,
                         {'step': '6', 'gid': '1'},
                         'group.volume_up')
    with pytest.raises(ValueError):
        await group.volume_up(0)
    with pytest.raises(ValueError):
        await group.volume_up(11)
    await group.volume_up(6)


@pytest.mark.asyncio
async def test_set_mute(mock_device, heos):
    """Test mute and unmute commands."""
    await heos.get_groups()
    group = heos.groups.get(1)

    mock_device.register(const.COMMAND_SET_GROUP_MUTE,
                         {"gid": "1", "state": "on"}, 'group.set_mute')
    await group.mute()

    mock_device.register(const.COMMAND_SET_GROUP_MUTE,
                         {"gid": "1", "state": "off"}, 'group.set_mute',
                         replace=True)
    await group.unmute()


@pytest.mark.asyncio
async def test_toggle_mute(mock_device, heos):
    """Test toggle mute command."""
    await heos.get_groups()
    group = heos.groups.get(1)
    mock_device.register(const.COMMAND_GROUP_TOGGLE_MUTE, {"gid": "1"},
                         'group.toggle_mute')
    await group.toggle_mute()
