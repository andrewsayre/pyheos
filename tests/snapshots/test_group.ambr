# serializer version: 1
# name: test_refresh
  dict({
    'group_id': -263109739,
    'is_muted': False,
    'lead_player_id': -263109739,
    'member_player_ids': list([
      845195621,
    ]),
    'name': 'Zone 1 + Zone 2',
    'volume': 42,
  })
# ---
# name: test_refresh_no_base_update
  dict({
    'group_id': 1,
    'is_muted': False,
    'lead_player_id': 1,
    'member_player_ids': list([
      2,
    ]),
    'name': 'Back Patio + Front Porch',
    'volume': 42,
  })
# ---
